Introduce edm::Async service, and use it in CUDA and Alpaka modules

makortel opened the pull request:
#### PR description:

This PR adds `edm::async()` facility described in https://github.com/cms-sw/cmssw/issues/29188, but implemented as a `edm::Async` service base class and `AsyncService` implementation (rationale for service-based approach instead of a free function can be found from https://github.com/cms-sw/cmssw/pull/44901#issuecomment-2103335066). This PR also replaces the use of `cudaStreamAddCallback()` with `edm::async()` accompanied with `cudaEventSynchronize()`, and makes the CUDA/Alpaka events to be created with `cudaEventBlockingSync` flag.

Measurements that I showed in CHEP 2023 https://indico.jlab.org/event/459/contributions/11810/ suggested possible 1 % throughput improvement at the HLT (of that time, many things have changed since) over `cudaStreamAddCallback()`. Earlier studies done with a prototype in https://github.com/cms-patatrack/pixeltrack-standalone/pull/321 that somehow the thread pool with `cudaEventSynchronize()` used less CPU than `cudaStreamAddCallback()`.

During the CHEP study I also tested polling with `cudaEventQuery()`, but the "waiting thread pool" approach was more performant.

Another benefit over `cudaStreamAddCallback()` is that that function "is slated for eventual deprecation and removal", and the "replacement" `cudaLaunchHostFunc()` does not call the callback function in case of an error in the associated CUDA stream.

Resolves https://github.com/cms-sw/cmssw/issues/29188
Resolves https://github.com/cms-sw/framework-team/issues/916

#### PR validation:

Unit tests in `FWCore/Concurrency`, `HeterogeneousCore/Alpaka{Core,Test}`, `HeterogeneousCore/CUDA{Utilities,Core,Test}` succeed.

**The deployment on CUDA and Alpaka modules still needs performance testing**

#### If this PR is a backport please specify the original PR and why you need to backport that PR. If this PR will be backported please specify to which release cycle the backport is meant for:

Possibly to be backported to 14_0_X.

makortel committed:
Add edm::WaitingThreadPool

makortel committed:
Use edm::Async service in CUDA modules

makortel committed:
Use edm::Async service in Alpaka modules

cmsbuild commented:
cms-bot internal usage<!-- bot cache: {"commits":{"23c43362779e89eab0e82d6339de9325f249ba0c":{"files":["FWCore/Concurrency/interface/async.h","FWCore/Concurrency/src/async.cc","FWCore/Concurrency/test/test_catch2_async.cc"],"squashed":false,"time":1714743349},"b758c1023837ef764d3044283907216b88c2f0e4":{"files":["HeterogeneousCore/AlpakaCore/interface/EventCache.h","HeterogeneousCore/AlpakaCore/interface/alpaka/EDMetadata.h","HeterogeneousCore/AlpakaCore/src/alpaka/EDMetadata.cc"],"squashed":false,"time":1714743350},"d96a64c59ec914476331e04bd6fc440f37bcb7e2":{"files":["HeterogeneousCore/CUDACore/src/ScopedContext.cc","HeterogeneousCore/CUDAUtilities/src/EventCache.cc"],"squashed":false,"time":1714743349}},"emoji":{},"last_seen_sha":"b758c1023837ef764d3044283907216b88c2f0e4","signatures":{}} -->

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-44901/40168

- This PR adds an extra 28KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  HeterogeneousCore/CUDACore/src/ScopedContext.cc  modified in PR(s): #37952

cmsbuild commented:
A new Pull Request was created by  @makortel for master.

It involves the following packages:

- FWCore/Concurrency (**core**)
- HeterogeneousCore/AlpakaCore (**heterogeneous**)
- HeterogeneousCore/CUDACore (**heterogeneous**)
- HeterogeneousCore/CUDAUtilities (**heterogeneous**)


@cmsbuild, @makortel, @Dr15Jones, @smuzaffar, @fwyzard can you please review it and eventually sign? Thanks.
@missirol, @wddgit, @rovere this is something you requested to watch as well.
@antoniovilela, @rappoccio, @sextonkennedy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

makortel commented:
@cmsbuild, please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d3cb6f/39230/summary.html
**COMMIT**: b758c1023837ef764d3044283907216b88c2f0e4
**CMSSW**: CMSSW_14_1_X_2024-05-03-1100/el8_amd64_gcc12
**Additional Tests**: GPU
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/44901/39230/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially [removed 4 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-05-03-1100+d3cb6f/62656/validateJR/logRootQA-events.log) from the logs
 - [Reco comparison results: 6 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-05-03-1100+d3cb6f/62656/validateJR.html)
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3331548
 - DQMHistoTests: [Total failures: 6](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-05-03-1100+d3cb6f/62656/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3331522
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 202 log files, 165 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-05-03-1100+d3cb6f/62656/triggerResults): no differences found


## GPU Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - [Reco comparison results: 0 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_14_1_X_2024-05-03-1100+d3cb6f/62657/validateJR.html)
 - DQMHistoTests: Total files compared: 3
 - DQMHistoTests: Total histograms compared: 39740
 - DQMHistoTests: [Total failures: 18](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_14_1_X_2024-05-03-1100+d3cb6f/62657/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 39722
 - DQMHistoTests: Total skipped: 0
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 2 files compared)
 - Checked 8 log files, 10 edm output root files, 3 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_14_1_X_2024-05-03-1100+d3cb6f/62657/triggerResults): no differences found

fwyzard commented on FWCore/Concurrency/src/async.cc:
```diff
@@ -0,0 +1,30 @@
+#include "FWCore/Concurrency/interface/async.h"
+
+namespace edm::impl {
+  WaitingThread::WaitingThread() { thread_ = std::thread(&WaitingThread::threadLoop, this); }
```
Could you set a meaningful thread name, so it's easier to identify this thread pool in a GDB trace ?

For example:
```suggestion
  WaitingThread::WaitingThread() {
    thread_ = std::thread(&WaitingThread::threadLoop, this);
    pthread_setname_np(thread_.native_handle(), "edm async pool");
  }
```

Or even something more elaborate with a `static constexpr` name, and a check that its length is 15 or less.

makortel replied:
Thanks for the suggestion, I added the name. While verifying the behavior by running the unit test in `gdb` I discovered many threads (~8) with the `edm async pool` name. Further investigation showed that the test ended up creating two `WaitingThread`s (which I can believe), but the use of `global_control` to set allowed parallelism to 1 lead the call to `onetbb::task_arena::enqueue()` in `edm::WaitingTaskWithArenaHolder::doneWaiting()` to create a new TBB-controlled thread that inherited the `edm async pool` name (and all subsequent TBB threads created by that thread inherited the name too).

When I set the allowed parallelism to 2 (which is now done in the test), I saw 2 threads with the `edm async pool` name.

While nearly all production jobs are configured to use multiple threads, maybe it would be time to look more into trying to give names to the TBB threads etc. I'll open a separate issue on that.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-44901/40191

- This PR adds an extra 20KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  HeterogeneousCore/CUDACore/src/ScopedContext.cc  modified in PR(s): #37952

cmsbuild commented:
Pull request #44901 was updated. @smuzaffar, @Dr15Jones, @fwyzard, @cmsbuild, @makortel can you please check and sign again.

makortel commented:
@cmsbuild, please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d3cb6f/39266/summary.html
**COMMIT**: f8322a08368fca17ce3c3ab9b58fecee94b1e1c0
**CMSSW**: CMSSW_14_1_X_2024-05-06-1100/el8_amd64_gcc12
**Additional Tests**: GPU
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/44901/39266/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially [added 3 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-05-06-1100+d3cb6f/62685/validateJR/logRootQA-events.log) to the logs
 - [Reco comparison results: 4 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-05-06-1100+d3cb6f/62685/validateJR.html)
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3332476
 - DQMHistoTests: [Total failures: 3](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-05-06-1100+d3cb6f/62685/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3332453
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 202 log files, 165 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-05-06-1100+d3cb6f/62685/triggerResults): no differences found


## GPU Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - [Reco comparison results: 0 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_14_1_X_2024-05-06-1100+d3cb6f/62686/validateJR.html)
 - DQMHistoTests: Total files compared: 3
 - DQMHistoTests: Total histograms compared: 39740
 - DQMHistoTests: [Total failures: 19](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_14_1_X_2024-05-06-1100+d3cb6f/62686/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 39721
 - DQMHistoTests: Total skipped: 0
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 2 files compared)
 - Checked 8 log files, 10 edm output root files, 3 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_14_1_X_2024-05-06-1100+d3cb6f/62686/triggerResults): no differences found

fwyzard commented:
I've compared some different approaches using the current HLT menu and recent data.

### `alpaka::HostOnlyTask`
Using `CMSSW_14_0_6_MULTIARCHS` as the baseline, without any changes, I get:
```
Running 4 times over 10300 events with 8 jobs, each with 32 threads, 24 streams and 1 GPUs
   692.8 ±   0.2 ev/s (10000 events, 98.2% overlap)
   691.2 ±   0.2 ev/s (10000 events, 98.3% overlap)
   692.4 ±   0.2 ev/s (10000 events, 98.2% overlap)
   690.7 ±   0.2 ev/s (10000 events, 98.9% overlap)
 --------------------
   691.8 ±   1.0 ev/s
```

### alpaka native host tasks
Using  `CMSSW_14_0_6_MULTIARCHS` with a modified `HeterogeneousCore/AlpakaCore/src/alpaka/EDMetadata.cc`:
```c++
  void EDMetadata::enqueueCallback(edm::WaitingTaskWithArenaHolder holder) {
    alpaka::enqueue(*queue_, [holder = std::move(holder)]() {
      // The functor is required to be const, but the original waitingTaskHolder_
      // needs to be notified...
      std::exception_ptr eptr;
      const_cast<edm::WaitingTaskWithArenaHolder&>(holder).doneWaiting(eptr);
    });
  }
```
I get:
```
Running 4 times over 10300 events with 8 jobs, each with 32 threads, 24 streams and 1 GPUs
   671.0 ±   0.2 ev/s (10000 events, 98.1% overlap)
   668.1 ±   0.2 ev/s (10000 events, 98.3% overlap)
   673.5 ±   0.2 ev/s (10000 events, 99.2% overlap)
   671.9 ±   0.2 ev/s (10000 events, 97.3% overlap)
 --------------------
   671.1 ±   2.3 ev/s
```

### `edm::async`
Using `CMSSW_14_0_6_MULTIARCHS` with the changes from this PR (as of f8322a08368fca17ce3c3ab9b58fecee94b1e1c0):
```
Running 4 times over 10300 events with 8 jobs, each with 32 threads, 24 streams and 1 GPUs
   703.8 ±   0.2 ev/s (10000 events, 97.2% overlap)
   703.5 ±   0.2 ev/s (10000 events, 98.8% overlap)
   700.3 ±   0.2 ev/s (10000 events, 98.4% overlap)
   701.1 ±   0.2 ev/s (10000 events, 99.0% overlap)
 --------------------
   702.2 ±   1.8 ev/s
```

---

Bottom line, this PR speeds up the HLT by about 1.5% ± 0.3% .

fwyzard commented:
+heterogeneous

makortel commented:
Notes from review discussion with @Dr15Jones 
- Have `edm::async()` to take the `WaitingThreadPool` as an argument
- Move the ownership of `WaitingThreadPool` object from a function-static (i.e. global) to a Service (that would be part of the default `cmsRun` services)
- Create a base class for the Service in order to decouple things. The base class would be placed in `FWCore/Concurrency`, and the Service implementation in `FWCore/Services`
- The Service base class would have a member function with similar interface as the `edm::async()` in this PR presently
- Additional benefits/opportunities of the Service approach
  - We can add ActivityRegistry signals around the caller-supplied functor (beyond this PR)
  - We can add caller-supplied context information in the exception message from the functor

Comments beyond this PR
- We should add a test where an exception (from the external worker) is passed to the `WaitingTaskWithArenaHolder::doneWaiting()`, and see what context is in the message

makortel commented on FWCore/Concurrency/src/async.cc:
```diff
@@ -0,0 +1,50 @@
+
+    do {
+      cond_.wait(lk, [this]() { return static_cast<bool>(func_) or stopThread_; });
+      if (func_) {
```
Check and act on `stopThread_` first (to avoid really strange failures)

makortel commented on FWCore/Concurrency/src/async.cc:
```diff
@@ -0,0 +1,50 @@
+    }
+  }
+
+  void WaitingThread::threadLoop() {
```
Mark this function as `noexcept`

makortel commented on FWCore/Concurrency/src/async.cc:
```diff
@@ -0,0 +1,50 @@
+
+    int err = pthread_setname_np(thread_.native_handle(), poolName);
+    // According to the glibc documentation, the only error
+    // pthread_setname_np() can returkn is about the argument C-string
```
```suggestion
    // pthread_setname_np() can return is about the argument C-string
```

makortel commented on FWCore/Concurrency/src/async.cc:
```diff
@@ -0,0 +1,50 @@
+    // When we are shutting down, we don't care about any possible
+    // system errors anymore
+    CMS_SA_ALLOW try {
+      if (not stopThread_) {
```
Remove the if, and make `stopThread()` function private.

makortel committed:
Add edm::Async service base class

makortel committed:
Add AsyncService service implementation

makortel committed:
Add AsyncService to the set of default services

makortel commented:
I addressed the comments in https://github.com/cms-sw/cmssw/pull/44901#issuecomment-2103335066 except for
> * We can add ActivityRegistry signals around the caller-supplied functor (beyond this PR)
> * We can limit the async launching to specific parts of the data processing (like only in module `acquire()` functions)

that I decided to leave for later PRs (and not to be backported to 14_0_X, unless requested).

Starting with this PR @Dr15Jones and I decided to start collecting generally useful testing modules to `FWCore/TestModules` (later on we are planning to move other modules there e.g. from `FWCore/Framework/test/stubs` and `FWCore/Integration`).

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-44901/40437

- This PR adds an extra 48KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  HeterogeneousCore/CUDACore/src/ScopedContext.cc  modified in PR(s): #37952

makortel commented:
@cmsbuild, please test

makortel commented:
@Dr15Jones Please review

makortel commented on FWCore/TestModules/plugins/StreamIDFilter.cc:
```diff
@@ -0,0 +1,35 @@
+  }
+}  // namespace edmtest
+
+DEFINE_FWK_MODULE(edmtest::StreamIDFilter);
```
@Dr15Jones What convention do we want here? I can think of
* `edmtest::<module>`
* `edm::<module>`
* Have the `<module>` defined in some namespace, but define it to the framework via `<module>` type alias (i.e. without a namespace)

I started with the `edmtest::<module>`, but at this stage it would be easy to change.

makortel replied:
Or maybe other way to phrase the question would be if `edmtest` would be appropriate namespace for such generally-usable modules, or do we want to preserve it to modules that are specific for testing the framework?

Dr15Jones replied:
I think we want to use the `edmtest` for any useful module that is meant for testing and NOT meant for use in production. I also think `edmtest` should appear in the python configurations as it clearly shows this is a test.

makortel replied:
Thanks, so we can continue with my present approach.

makortel commented:
On another look, I think I'm still missing

> * We can destruct the `WaitingThreadPool` at the endJob, or Service destructor, instead of "after `main()`"

I'll look into it.

cmsbuild commented:
-1

**Failed Tests**: UnitTests GpuUnitTests 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d3cb6f/39635/summary.html
**COMMIT**: 6c12a5884cc2f120682e44fc1bafdfe21175ca58
**CMSSW**: CMSSW_14_1_X_2024-05-30-1100/el8_amd64_gcc12
**Additional Tests**: GPU
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/44901/39635/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Unit Tests

I found 2 errors in the following unit tests:

<pre>---> test TestFWCoreServicesDriver_AsyncService had ERRORS
---> test RecoPPSLocalNewT2 had ERRORS
</pre>

## GPU Unit Tests

I found 1 errors in the following unit tests:

<pre>---> test testHeterogeneousCoreCUDACore had ERRORS
</pre>

## Comparison Summary



**Summary**:
 - You potentially [added 1 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-05-30-1100+d3cb6f/62993/validateJR/logRootQA-events.log) to the logs
 - [Reco comparison results: 4 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-05-30-1100+d3cb6f/62993/validateJR.html)
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3338862
 - DQMHistoTests: [Total failures: 3](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-05-30-1100+d3cb6f/62993/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3338839
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 202 log files, 165 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-05-30-1100+d3cb6f/62993/triggerResults): no differences found


## GPU Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - [Reco comparison results: 0 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_14_1_X_2024-05-30-1100+d3cb6f/62992/validateJR.html)
 - DQMHistoTests: Total files compared: 3
 - DQMHistoTests: Total histograms compared: 39740
 - DQMHistoTests: [Total failures: 23](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_14_1_X_2024-05-30-1100+d3cb6f/62992/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 39717
 - DQMHistoTests: Total skipped: 0
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 2 files compared)
 - Checked 8 log files, 10 edm output root files, 3 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_14_1_X_2024-05-30-1100+d3cb6f/62992/triggerResults): no differences found

Dr15Jones commented on FWCore/Concurrency/interface/WaitingThreadPool.h:
```diff
@@ -0,0 +1,106 @@
+      void run(WaitingTaskWithArenaHolder holder,
+               F&& func,
+               G&& errorContextFunc,
+               std::shared_ptr<WaitingThread> thisPtr) {
```
Should this be `noexcept`?

makortel replied:
Depends if we want possible system errors from locking the mutex to be reported via exceptions to be caught by the caller, or via `std::terminate()`.

makortel replied:
Let's leave as it is.

Dr15Jones commented on FWCore/Concurrency/interface/WaitingThreadPool.h:
```diff
@@ -0,0 +1,106 @@
+      }
+
+    private:
+      void stopThread() {
```
Could be `noexcept`

makortel replied:
Let's leave as it is.

Dr15Jones commented on FWCore/Concurrency/interface/Async.h:
```diff
@@ -0,0 +1,28 @@
+  class Async {
+  public:
+    Async() = default;
+    virtual ~Async();
```
Should this be marked `noexcept`?

makortel replied:
It would probably be good. Looking what gets done by the destructor, presently the destructor of `edm::ReusableObjectHolder` is not `noexcept`. Could or should it be?
 https://github.com/cms-sw/cmssw/blob/774f85421329c42cd1d30ec43d470f527141fb92/FWCore/Utilities/interface/ReusableObjectHolder.h#L91-L97
The OneAPI spec doesn't say anything about exceptions in `oneapi::tbb::concurrent_queue<...>::try_pop()` https://spec.oneapi.io/versions/latest/elements/oneTBB/source/containers/concurrent_queue_cls/safe_member_functions.html#popping-elements
so I guess the question is what we want the behavior to be if the destructor of `T` throws.

Well, if the destructor of `T` would throw, the `while` loop would terminate early and leave the remaining `T` to not be destructed, which is probably bad enough to just call `std::terminate` (i.e. suggesting to mark the `ReusableObjectHolder` destructor as `noexcept`.

Dr15Jones commented on FWCore/Concurrency/test/test_catch2_WaitingThreadPool.cc:
```diff
@@ -0,0 +1,258 @@
+      using namespace edm::waiting_task::chain;
+      auto h = first([&pool, &count](edm::WaitingTaskHolder h) {
+                 edm::WaitingTaskWithArenaHolder h2(std::move(h));
+                 pool.runAsync(
```
What is the underlying behavior you are trying to test here?

makortel replied:
The system works also when there are multiple `runAsync()` calls from the same task (i.e. both functions have finished at the time `FinalWaitingTask::waitNoThrow()` returns).

Dr15Jones commented on FWCore/Concurrency/test/test_catch2_WaitingThreadPool.cc:
```diff
@@ -0,0 +1,258 @@
+                  pool.runAsync(
+                      h2, [&count]() { ++count; }, errorContext);
+                }) |
+                lastTask(edm::WaitingTaskHolder(group, &waitTask));
```
How do you mean this test to be different from the one above?

makortel replied:
My intention here is to have `runAsync()` called concurrently (from two different tasks).

makortel replied:
Add spinloops such that there would be an attempt to call `pool.runAsync()` simultaneously.

Dr15Jones commented on FWCore/Concurrency/test/test_catch2_WaitingThreadPool.cc:
```diff
@@ -0,0 +1,258 @@
+        h.doneWaiting(std::exception_ptr());
+      }
+      REQUIRE_THROWS_AS(waitTask.wait(), cms::Exception);
+      REQUIRE(count.load() == 1);
```
Is this actually guaranteed to work? Couldn't `"1"` thread start and pause at line 134 then `"2"` thread finishes and only after that `"1"` continues? In that case `count.load() == 2`.

makortel replied:
Good point, thanks. I need to use `fetch_add()`.

Dr15Jones commented on FWCore/Services/plugins/AsyncService.cc:
```diff
@@ -0,0 +1,47 @@
+  AsyncService::AsyncService(ParameterSet const& iConfig, ActivityRegistry& iRegistry) {
+    iRegistry.watchPreSourceEarlyTermination([this](TerminationOrigin) { allowed_ = false; });
+    iRegistry.watchPreGlobalEarlyTermination([this](GlobalContext const&, TerminationOrigin) { allowed_ = false; });
+    iRegistry.watchPreStreamEarlyTermination([this](StreamContext const&, TerminationOrigin) { allowed_ = false; });
```
Should we also add `postEndJob`?

makortel replied:
I see it would be consistent. I could do it already in `preEndJob`, but then the exception message should be improved, and in a way that would be step towards restricting more when the `AsyncService` can be used that I decided to leave for a later PR. I'll add the `postEndJob`.

makortel commented:
This update should fix the failed unit tests (last part of `TestFWCoreServicesDriver_AsyncService` turned out to be flawed, so that lead to a big larger reorganization)

makortel commented:
@cmsbuild, please test

To see the tests actually got fixed before proceeding to @Dr15Jones' comments

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-44901/40451

- This PR adds an extra 24KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  HeterogeneousCore/CUDACore/src/ScopedContext.cc  modified in PR(s): #37952

cmsbuild commented:
Pull request #44901 was updated. @Dr15Jones, @fwyzard, @smuzaffar, @makortel can you please check and sign again.

makortel commented on FWCore/Concurrency/interface/Async.h:
```diff
@@ -0,0 +1,28 @@
+  public:
+    Async() = default;
+    virtual ~Async();
+
```
The `WaitingThreadPool` member already prohibits copying and moving, but it's probably good to be explicit here.

makortel commented on FWCore/Concurrency/src/WaitingThreadPool.cc:
```diff
@@ -0,0 +1,55 @@
+    while (true) {
+      cond_.wait(lk, [this]() { return static_cast<bool>(func_) or stopThread_; });
+      if (stopThread_) {
+        break;
```
The question came to my mind as well, so I'll add the assert.

makortel committed:
Mark ReusableObjectHolder destructor as noexcept

Pretty much only piece that could throw is the destructor of T, and
bad things would happen if the while loop would terminate early, so we
could just call std::terminate() in case of an exception.

cmsbuild commented:
-1

**Failed Tests**: UnitTests 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d3cb6f/39651/summary.html
**COMMIT**: 0c9b89efa8249d4ed86a7e6e905b74ce413aa392
**CMSSW**: CMSSW_14_1_X_2024-05-31-1100/el8_amd64_gcc12
**Additional Tests**: GPU
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/44901/39651/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Unit Tests

I found 1 errors in the following unit tests:

<pre>---> test RecoPPSLocalNewT2 had ERRORS
</pre>

## Comparison Summary



There are some workflows for which there are errors in the baseline:
 1000.0 step 2
1001.0 step 2
101.0 step 1
10224.0 step 1
11634.0 step 1
12434.0 step 1
12834.0 step 1
12834.7 step 1
12846.0 step 1
13034.0 step 1
1306.0 step 1
13234.0 step 1
1330.0 step 1
135.4 step 1
136.731 step 2
136.7611 step 2
136.793 step 2
136.8311 step 2
136.874 step 2
136.88811 step 2
139.001 step 2
140.023 step 2
140.043 step 2
140.063 step 2
140.56 step 2
14034.0 step 1
141.042 step 2
141.044 step 2
141.046 step 2
14234.0 step 1
158.01 step 2
23234.0 step 1
24834.0 step 1
24834.911 step 1
24896.0 step 1
24900.0 step 1
25.0 step 1
2500.4 step 2
250202.181 step 1
25034.999 step 1
25202.0 step 1
312.0 step 1
4.22 step 2
4.53 step 2
5.1 step 1
7.3 step 1
8.0 step 1
9.0 step 1
 The results for the comparisons for these workflows could be incomplete 
This means most likely that the IB is having errors in the relvals.The error does NOT come from this pull request

**Summary**:
 - You potentially [added 19180 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-05-31-1100+d3cb6f/63004/validateJR/logRootQA-events.log) to the logs
 - [Reco comparison results: 0 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-05-31-1100+d3cb6f/63004/validateJR.html)
 - DQMHistoTests: Total files compared: 1
 - DQMHistoTests: Total histograms compared: 0
 - DQMHistoTests: [Total failures: 0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-05-31-1100+d3cb6f/63004/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 0
 - DQMHistoTests: Total skipped: 0
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0 KiB( 0 files compared)
 - Checked 66 log files, 0 edm output root files, 1 DQM output files


## GPU Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - [Reco comparison results: 48 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_14_1_X_2024-05-31-1100+d3cb6f/63003/validateJR.html)
 - DQMHistoTests: Total files compared: 3
 - DQMHistoTests: Total histograms compared: 39740
 - DQMHistoTests: [Total failures: 1783](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_14_1_X_2024-05-31-1100+d3cb6f/63003/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 37957
 - DQMHistoTests: Total skipped: 0
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 2 files compared)
 - Checked 8 log files, 10 edm output root files, 3 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_14_1_X_2024-05-31-1100+d3cb6f/63003/triggerResults): no differences found

makortel committed:
Spinloops in WaitingThreadPool test

makortel commented:
Addressed the review comments from @Dr15Jones

makortel commented:
@cmsbuild, please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-44901/40480

- This PR adds an extra 56KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  HeterogeneousCore/CUDACore/src/ScopedContext.cc  modified in PR(s): #37952

cmsbuild commented:
Pull request #44901 was updated. @fwyzard, @Dr15Jones, @makortel, @smuzaffar can you please check and sign again.

cmsbuild commented:
-1

**Failed Tests**: UnitTests 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d3cb6f/39706/summary.html
**COMMIT**: ace8492258c0b90c417da8159355188797d4c53b
**CMSSW**: CMSSW_14_1_X_2024-06-04-2300/el8_amd64_gcc12
**Additional Tests**: GPU
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/44901/39706/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Unit Tests

I found 1 errors in the following unit tests:

<pre>---> test RecoPPSLocalNewT2 had ERRORS
</pre>

## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - [Reco comparison results: 11 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-06-04-2300+d3cb6f/63050/validateJR.html)
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3338862
 - DQMHistoTests: [Total failures: 9](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-06-04-2300+d3cb6f/63050/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3338833
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 202 log files, 165 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-06-04-2300+d3cb6f/63050/triggerResults): no differences found


## GPU Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - [Reco comparison results: 0 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_14_1_X_2024-06-04-2300+d3cb6f/63049/validateJR.html)
 - DQMHistoTests: Total files compared: 3
 - DQMHistoTests: Total histograms compared: 39740
 - DQMHistoTests: [Total failures: 22](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_14_1_X_2024-06-04-2300+d3cb6f/63049/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 39718
 - DQMHistoTests: Total skipped: 0
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 2 files compared)
 - Checked 8 log files, 10 edm output root files, 3 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_14_1_X_2024-06-04-2300+d3cb6f/63049/triggerResults): no differences found

makortel commented:
Unit test fails in IBs https://github.com/cms-sw/cmssw/issues/45101

makortel commented:
+core

makortel commented:
+heterogeneous

Based on earlier https://github.com/cms-sw/cmssw/pull/44901#issuecomment-2099490033, and that on `heterogeneous` side there were no substantial changes.

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (but tests are reportedly failing). This pull request will now be reviewed by the release team before it's merged. @antoniovilela, @sextonkennedy, @rappoccio (and backports should be raised in the release meeting by the corresponding L2)

makortel commented:
ignore tests-rejected with ib-failure

rappoccio commented:
+1

rappoccio commented:
merge
